### DISCUSSION TITLE
[@types/react] `forwardRef` method return statement allows string references

### DISCUSSION THREAD

**Discussion by bsunderhus** [DefinitelyTyped/DefinitelyTyped#69756] [Issues with a @types package] [answered]
String refs are only supported by Class components as per React documentation https://react.dev/blog/2024/04/25/react-19-upgrade-guide#removed-string-refs

At the moment the new `@types/react@18` introduced  `LegacyRef` to `RefAttributes` and `forwardRef` components now are allowing the usage of `string` as references, which is an error and will throw if you try to do so: https://stackblitz.com/edit/vitejs-vite-vhvsvw?file=src%2Fmain.tsx

## Proposed change

I propose that instead of using `RefAttributes`, `forwardRef` should declare it's reference value directly!

```ts
// right now 💣
function forwardRef<T, P = {}>(
    render: ForwardRefRenderFunction<T, P>,
): ForwardRefExoticComponent<PropsWithoutRef<P> & RefAttributes<T>>;

// solution ✅
function forwardRef<T, P = {}>(
    render: ForwardRefRenderFunction<T, P>,
): ForwardRefExoticComponent<PropsWithoutRef<P> & {ref?: Ref<T>}>;
```

### Alternative proposal

Just rollback to the way it was on v17: 

```ts
/**
 * @internal You shouldn't need to use this type since you never see these attributes
 * inside your component or have to validate them.
 */
interface Attributes {
    key?: Key | null | undefined;
}
interface RefAttributes<T> extends Attributes {
    ref?: Ref<T> | undefined;
}
interface ClassAttributes<T> extends Attributes {
    ref?: LegacyRef<T> | undefined;
}
```

[a comment by an automated account is not shown]

**Comment by eps1lon** [comment, marked as answer]
> String refs are only supported by Class components as per React documentation

They are supported **in** class components which the error mentions:
> You may be adding a ref to a component that was not created inside a component's render method

 What you pass the ref to can be just any component that accepts refs: class components, built-in components (e.g. `div`, `button` etc) and `forwardRef` components (and function components in React 19).

What you showed, would've also errored with class components and built-in components. 

The following sandbox showcases how string refs on `forwardRef` **inside** class components is fine while string refs error on **any** component if used outside of a class component: https://stackblitz.com/edit/vitejs-vite-slejom?file=src%2Fmain.tsx
